### PR TITLE
feat(coderd): add matched provisioner daemons information to more places

### PR DESCRIPTION
**Pull request by johnstcn**
- [x] Refactors `checkProvisioners` into `db2sdk.MatchedProvisioners`
- [x] Adds a separate RBAC subject just for reading provisioner daemons for this use-case
- [x] Adds matched provisioners information to more templateversion endpoints
- [x] Updates existing unit tests for template version endpoints
- [x] Updates `wsbuilder` to also return matched provisioners
- [x] Updates existing unit tests for workspaces API endpoints
- [x] Updates CLI to show a warning when creating a workspace for which no provisoners are available
- [x] Adds API endpoint for matched provisioners of template dry-run job
- [x] Local smoke-testing (stop and start external provisioner daemon while testing tagged templates/workspaces)

### CODE REVIEW

File: coderd/templateversions.go [hunk 1]
```diff
@@ -1822,34 +1912,3 @@ func (api *API) publishTemplateUpdate(ctx context.Context, templateID uuid.UUID)
...
-		if provisioner.LastSeenAt.Time.After(staleInterval) {
-			matched.Available++
-		}
-		if provisioner.LastSeenAt.Time.After(mostRecentlySeen.Time) {
-			matched.MostRecentlySeen.Valid = true
-			matched.MostRecentlySeen.Time = provisioner.LastSeenAt.Time
-		}
-	}
-	return matched, nil
-}
```
Comment (johnstcn) [inline coderd/templateversions.go:1855]: self-review: extracted to `db2sdk`.

File: coderd/database/dbauthz/dbauthz.go [hunk 2]
```diff
@@ -359,6 +376,12 @@ func AsSystemRestricted(ctx context.Context) context.Context {
 	return context.WithValue(ctx, authContextKey{}, subjectSystemRestricted)
 }
 
+// AsSystemReadProvisionerDaemons returns a context with an actor that has permissions
+// to read provisioner daemons.
+func AsSystemReadProvisionerDaemons(ctx context.Context) context.Context {
+	return context.WithValue(ctx, authContextKey{}, subjectSystemReadProvisionerDaemons)
+}
+
```
Comment (johnstcn) [inline coderd/database/dbauthz/dbauthz.go:384]: self-review: I wanted to avoid sprinkling `dbauthz.AsSystemRestricted` everywhere, so I made a separate RBAC role for when we just wish to read provisioner daemons. I can remove this and switch back to `SystemRestricted` if folks prefer.

File: coderd/autobuild/lifecycle_executor.go [hunk 3]
```diff
@@ -245,7 +245,7 @@ func (e *Executor) runOnce(t time.Time) Stats {
 							}
 						}
 
-						nextBuild, job, err = builder.Build(e.ctx, tx, nil, audit.WorkspaceBuildBaggage{IP: "127.0.0.1"})
+						nextBuild, job, _, err = builder.Build(e.ctx, tx, nil, audit.WorkspaceBuildBaggage{IP: "127.0.0.1"})
```
Comment (johnstcn) [inline coderd/autobuild/lifecycle_executor.go:248]: self-review: we may need to notify or log about this, but deferring for later.

File: coderd/templateversions.go [hunk 4]
```diff
@@ -934,7 +983,23 @@ func (api *API) templateVersionByOrganizationTemplateAndName(rw http.ResponseWri
...
+			WantTags:       jobs[0].ProvisionerJob.Tags,
+		})
+		if err != nil {
+			api.Logger.Error(ctx, "failed to fetch provisioners for job id", slog.F("job_id", jobs[0].ProvisionerJob.ID), slog.Error(err))
+		} else {
+			matchedProvisioners = ptr.Ref(db2sdk.MatchedProvisioners(provisioners, dbtime.Now(), provisionerdserver.StaleInterval))
+		}
+	}
+
+	httpapi.Write(ctx, rw, http.StatusOK, convertTemplateVersion(templateVersion, convertProvisionerJob(jobs[0]), matchedProvisioners, nil))
```
Comment (johnstcn) [inline coderd/templateversions.go:1002]: self-review: This addresses an issue I noticed in the frontend where it quickly "flashes" with the tag warning message but then gets overwritten when the FE re-requests the template version. Adding it to other template-related endpoints for posterity.

File: coderd/templateversions_test.go [hunk 5]
```diff
@@ -50,6 +50,12 @@ func TestTemplateVersion(t *testing.T) {
 		tv, err := client.TemplateVersion(ctx, version.ID)
 		authz.AssertChecked(t, policy.ActionRead, tv)
 		require.NoError(t, err)
+		if assert.Equal(t, tv.Job.Status, codersdk.ProvisionerJobPending) {
+			assert.NotNil(t, tv.MatchedProvisioners)
+			assert.Zero(t, tv.MatchedProvisioners.Available)
+			assert.Zero(t, tv.MatchedProvisioners.Count)
+			assert.False(t, tv.MatchedProvisioners.MostRecentlySeen.Valid)
+		}
```
Comment (johnstcn) [inline coderd/templateversions_test.go:58]: zero because includeProvisionerDaemon is not set

File: coderd/templateversions_test.go [hunk 6]
```diff
@@ -158,6 +170,12 @@ func TestPostTemplateVersionsByOrganization(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, "bananas", version.Name)
 		require.Equal(t, provisionersdk.ScopeOrganization, version.Job.Tags[provisionersdk.TagScope])
+		if assert.Equal(t, version.Job.Status, codersdk.ProvisionerJobPending) {
+			assert.NotNil(t, version.MatchedProvisioners)
+			assert.Equal(t, version.MatchedProvisioners.Available, 1)
+			assert.Equal(t, version.MatchedProvisioners.Count, 1)
+			assert.True(t, version.MatchedProvisioners.MostRecentlySeen.Valid)
+		}
```
Comment (johnstcn) [inline coderd/templateversions_test.go:178]: includeProvisionerDaemon is set here!

File: coderd/wsbuilder/wsbuilder_test.go [hunk 7]
```diff
@@ -94,7 +95,8 @@ func TestBuilder_NoOptions(t *testing.T) {
 
 	ws := database.Workspace{ID: workspaceID, TemplateID: templateID, OwnerID: userID}
 	uut := wsbuilder.New(ws, database.WorkspaceTransitionStart)
-	_, _, err := uut.Build(ctx, mDB, nil, audit.WorkspaceBuildBaggage{})
+	// nolint: dogsled
```
Comment (johnstcn) [inline coderd/wsbuilder/wsbuilder_test.go:98]: self-review: we should probably disable this linter for unit tests

File: enterprise/coderd/workspacebuilds_test.go [hunk 8]
```diff
@@ -109,7 +109,7 @@ func TestWorkspaceBuild(t *testing.T) {
 
 		for _, c := range cases {
 			t.Run(c.Name, func(t *testing.T) {
-				_, err = c.Client.CreateWorkspace(ctx, owner.OrganizationID, codersdk.Me, codersdk.CreateWorkspaceRequest{
+				_, err = c.Client.CreateUserWorkspace(ctx, codersdk.Me, codersdk.CreateWorkspaceRequest{
```
Comment (johnstcn) [inline enterprise/coderd/workspacebuilds_test.go:112]: self-review: noticed this was usage of a deprecated function, drive-by fix.

File: coderd/workspacebuilds.go [hunk 9]
```diff
@@ -384,12 +387,18 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
...
-	err = provisionerjobs.PostJob(api.Pubsub, *provisionerJob)
-	if err != nil {
-		// Client probably doesn't care about this error, so just log it.
-		api.Logger.Error(ctx, "failed to post provisioner job to pubsub", slog.Error(err))
+
+	if provisionerJob != nil {
+		if err := provisionerjobs.PostJob(api.Pubsub, *provisionerJob); err != nil {
+			// Client probably doesn't care about this error, so just log it.
+			api.Logger.Error(ctx, "failed to post provisioner job to pubsub", slog.Error(err))
+		}
```
Comment (johnstcn) [inline coderd/workspacebuilds.go:395]: self-review: linter complained about a possible nil ptr reference.

File: coderd/workspacebuilds_test.go [hunk 10]
```diff
@@ -1181,6 +1205,102 @@ func TestPostWorkspaceBuild(t *testing.T) {
...
+		coderdtest.AwaitTemplateVersionJobCompleted(t, client, version.ID)
+		template := coderdtest.CreateTemplate(t, client, user.OrganizationID, version.ID)
+		workspace := coderdtest.CreateWorkspace(t, client, template.ID)
+		coderdtest.AwaitWorkspaceBuildJobCompleted(t, client, workspace.LatestBuild.ID)
+
+		// Stop the provisioner daemon.
+		require.NoError(t, closeDaemon.Close())
+		ctx := testutil.Context(t, testutil.WaitLong)
+		// Given: no provisioner daemons exist.
+		_, err := db.ExecContext(ctx, `DELETE FROM provisioner_daemons;`)
```
Comment (johnstcn) [inline coderd/workspacebuilds_test.go:1234]: self-review: this is kinda yucky but I don't agree with the idea of creating database queries for things that will only ever get used intests.

File: coderd/database/dbauthz/dbauthz.go [hunk 11]
```diff
@@ -317,6 +317,23 @@ var (
 		}),
 		Scope: rbac.ScopeAll,
 	}.WithCachedASTValue()
+
+	subjectSystemReadProvisionerDaemons = rbac.Subject{
+		FriendlyName: "System",
```
Comment (dannykopping) [inline coderd/database/dbauthz/dbauthz.go:322]: Nit

```suggestion
		FriendlyName: "Provisioner Daemons Reader",
```

File: coderd/templateversions.go [hunk 12]
```diff
@@ -60,6 +62,22 @@ func (api *API) templateVersion(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	var matchedProvisioners *codersdk.MatchedProvisioners
+	if jobs[0].ProvisionerJob.JobStatus == database.ProvisionerJobStatusPending {
```
Comment (dannykopping) [inline coderd/templateversions.go:66]: Do we only ever expect a single job to be returned by `GetProvisionerJobsByIDsWithQueuePosition`?

**Review by dannykopping — APPROVED** [review]
:shipit:

File: coderd/templateversions.go (hunk 12 above)
Comment (johnstcn) [reply coderd/templateversions.go:66]: In this instance, we should get either 0 or 1:
```
jobs, err := api.Database.GetProvisionerJobsByIDsWithQueuePosition(ctx, []uuid.UUID{templateVersion.JobID})
```

Comment (dannykopping) [reply coderd/templateversions.go:66]: My point here is really that if we're only ever expecting a single job, we should consider changing the semantics of the `GetProvisionerJobsByIDsWithQueuePosition` to be a `:one` not a `:many`.

Comment (dannykopping) [reply coderd/templateversions.go:66]: I guess it's _possible_ though that a single template version could have multiple provisioner jobs associated if something went wrong.
